lightning/src/chain/chainmonitor.rs

```diff
@@ -82,24 +83,82 @@ where C::Target: chain::Filter,
 	/// descendants of such transactions. It is not necessary to re-fetch the block to obtain
 	/// updated `txdata`.
 	pub fn block_connected(&self, header: &BlockHeader, txdata: &TransactionData, height: u32) {
+		self.process_transactions(header, txdata, |monitor, txdata| {
+			monitor.block_connected(
+				header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger)
+		});
+	}
+
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel and reacting accordingly to newly confirmed transactions. For details, see
+	/// [`ChannelMonitor::transactions_confirmed`].
+	///
+	/// Used instead of [`block_connected`] by clients that are notified of transactions rather than
+	/// blocks. May be called before or after [`update_best_block`] for transactions in the
+	/// corresponding block. See [`update_best_block`] for further calling expectations. 
+	///
+	/// [`block_connected`]: Self::block_connected
+	/// [`update_best_block`]: Self::update_best_block
+	pub fn transactions_confirmed(&self, header: &BlockHeader, txdata: &TransactionData, height: u32) {
+		self.process_transactions(header, txdata, |monitor, txdata| {
+			monitor.transactions_confirmed(
+				header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger)
+		});
+	}
+
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel and reacting accordingly based on the new chain tip. For details, see
+	/// [`ChannelMonitor::update_best_block`].
+	///
+	/// Used instead of [`block_connected`] by clients that are notified of transactions rather than
+	/// blocks. May be called before or after [`transactions_confirmed`] for the corresponding
+	/// block.
+	///
+	/// Must be called after new blocks become available for the most recent block. Intermediary
+	/// blocks, however, may be safely skipped. For a chain reorganization, must either be:
+	/// * called at at the fork point and then again for the most recent block -- intermediary
+	///   blocks may be skipped as before -- or
+	/// * not called in favor of calling [`transaction_unconfirmed`] for any affected transactions.
```

Comment:
> A bit confused by this alternative. Both apply to non-`block_connected` clients ?

Correct.

> But do we have such clients not being notified of block disconnections and only of transactions ones in such a way they're forced to use `transaction_unconfirmed` ?
> 

IIUC, Electrum clients will be able to tell which transactions were unconfirmed but may need to do bit more work to figure out block disconnects. @TheBlueMatt Perhaps you could clear this up. Should I just drop the fork point requirement here?

> Otherwise, I would favor an unified API here, unless there is another rational ?

Either way, it looks like the fork point requirement will go away in #853 according to https://github.com/rust-bitcoin/rust-lightning/pull/858#discussion_r611090222. So I can clean up the requirements when documenting the new trait in a follow-up PR.

Reply:
> IIUC, Electrum clients will be able to tell which transactions were unconfirmed but may need to do bit more work to figure out block disconnects. @TheBlueMatt Perhaps you could clear this up. Should I just drop the fork point requirement here?

Hmm, good question, I guess it depends on if we have any immediate users who would want to do that. Post-#843 the "call at the fork point" method should still work on `ChannelManager` as well, so we could leave it as an option if someone wanted to use it, but its unclear to me exactly if someone would, or whether they'd just be using the block_connected/block_disconnected APIs if they were ever doing chain tracking. I'm definitely fine with either dropping it or keeping it.